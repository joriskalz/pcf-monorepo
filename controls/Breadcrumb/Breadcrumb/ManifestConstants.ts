export const enum ManifestPropertyNames {
    dataset = 'dataset',
}

export const enum ItemColumns {
    DisplayName = 'ItemDisplayName',
    Key = 'ItemKey',
    Clickable = 'ItemClickable',
}

export const enum InputEvents {
    SetFocus = 'SetFocus',
}

export const enum InputProperties {
    InputEvent = 'InputEvent',
    SelectedKey = 'SelectedKey',
}
