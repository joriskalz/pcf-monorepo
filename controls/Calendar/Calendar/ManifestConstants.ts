export const enum InputEvents {
    SetFocus = 'SetFocus',
}
