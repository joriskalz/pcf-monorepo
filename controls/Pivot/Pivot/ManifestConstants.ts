export const enum ManifestPropertyNames {
    dataset = 'dataset',
}

export const enum ItemColumns {
    DisplayName = 'ItemDisplayName',
    Key = 'ItemKey',
    Enabled = 'ItemEnabled',
    IconName = 'ItemIconName',
    IconColor = 'ItemIconColor',
    TextColor = 'ItemTextColor',
    ParentKey = 'ItemParentKey',
    IconOnly = 'ItemIconOnly',
    Visible = 'ItemVisible',
    Checked = 'ItemChecked',
    ItemCount = 'ItemCount',
}

export const enum InputEvents {
    SetFocus = 'SetFocus',
    Reset = 'Reset',
}

export const enum InputProperties {
    SelectedKey = 'SelectedKey',
}

export class RenderTypes {
    static PivotLinks = '0';
    static PivotTabs = '1';
}

export class RenderSize {
    static Normal = '0';
    static Compact = '1';
    static Large = '2';
}
