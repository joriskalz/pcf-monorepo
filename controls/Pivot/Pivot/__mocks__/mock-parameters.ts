/* istanbul ignore file */

import { IInputs } from '../generated/ManifestTypes';
import { MockEnumProperty, MockStringProperty } from './mock-context';
import { MockDataSet } from './mock-datasets';

export function getMockParameters(): IInputs {
    return {
        AccessibilityLabel: new MockStringProperty(),
        InputEvent: new MockStringProperty(),
        RenderSize: new MockEnumProperty(),
        RenderType: new MockEnumProperty(),
        SelectedKey: new MockStringProperty(),
        Theme: new MockStringProperty(),
        items: new MockDataSet([]),
    };
}
