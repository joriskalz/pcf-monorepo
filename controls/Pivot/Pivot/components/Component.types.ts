export interface PivotLink {
    id: string;
    enabled?: boolean;
    name: string;
    iconName?: string;
    iconColor?: string;
    backgroundColor?: string;
    borderColor?: string;
    textColor?: string;
    itemCount?: number;
    visible?: boolean;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    data: any;
    key: string;
    checked: boolean;
}

export interface PivotProps {
    width?: number;
    height?: number;
    items: PivotLink[];
    onSelected: (item: PivotLink) => void;
    disabled?: boolean;
    tabIndex?: number;
    ariaLabel?: string;
    themeJSON?: string;
    selectedKey?: string;
    setFocus?: string;
    linkSize: 'normal' | 'large';
    linkFormat: 'links' | 'tabs';
}
