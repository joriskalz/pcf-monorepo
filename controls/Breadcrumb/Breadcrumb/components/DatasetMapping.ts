import { IBreadcrumbItem } from '@fluentui/react';
import { ItemColumns } from '../ManifestConstants';
import { ICustomBreadcrumbItem } from './components.types';

export function getItemsFromDataset(dataset: ComponentFramework.PropertyTypes.DataSet): ICustomBreadcrumbItem[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return [getDummyAction('1'), getDummyAction('2'), getDummyAction('3')];
    }
    const keyIndex: Record<string, number> = {};
    return dataset.sortedRecordIds.map((id) => {
        const record = dataset.records[id];
        // Prevent duplicate keys by appending the duplicate index
        let key = record.getValue(ItemColumns.Key) as string;
        if (keyIndex[key] !== undefined) {
            keyIndex[key]++;
            key += `_${keyIndex[key]}`;
        } else keyIndex[key] = 1;
        return {
            id: record.getRecordId(),
            key: key,
            text: record.getFormattedValue(ItemColumns.DisplayName),
            clickable: record.getValue(ItemColumns.Clickable) as boolean,
            data: record,
        } as ICustomBreadcrumbItem;
    });
}

export function getBreadcrumbItems(
    items: ICustomBreadcrumbItem[],
    onClick: (ev?: React.MouseEvent<HTMLElement>, item?: ICustomBreadcrumbItem) => void,
): IBreadcrumbItem[] {
    return items.map((item) =>
        item.clickable === false
            ? {
                  id: item.id,
                  key: item.key,
                  text: item.text,
              }
            : {
                  id: item.id,
                  key: item.key,
                  text: item.text,
                  onClick: onClick,
              },
    ) as IBreadcrumbItem[];
}

function getDummyAction(key: string): ICustomBreadcrumbItem {
    return {
        id: key,
        key: key,
        text: 'Item ' + key,
        clickable: true,
    } as ICustomBreadcrumbItem;
}
