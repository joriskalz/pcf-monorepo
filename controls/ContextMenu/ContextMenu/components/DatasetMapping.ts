import { CanvasContextMenuItem } from './Component.types';
import {
    ContextualMenuItemType,
    getColorFromString,
    getShade,
    IButtonStyles,
    IContextualMenuItem,
    IIconProps,
    Shade,
} from '@fluentui/react';
import { ItemColumns } from '../ManifestConstants';
const CSS_IMPORTANT = ' !important';
export function getMenuItemsFromDataset(dataset: ComponentFramework.PropertyTypes.DataSet): CanvasContextMenuItem[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return [getDummyAction('1'), getDummyAction('2'), getDummyAction('3')];
    }
    const keyIndex: Record<string, number> = {};
    return dataset.sortedRecordIds.map((id) => {
        const record = dataset.records[id];
        // Prevent duplicate keys by appending the duplicate index
        let key = (record.getFormattedValue(ItemColumns.Key) as string) ?? '';
        if (keyIndex[key] !== undefined) {
            keyIndex[key]++;
            key += `_${keyIndex[key]}`;
        } else keyIndex[key] = 1;
        return {
            id: record.getRecordId(),
            key: key,
            name: record.getFormattedValue(ItemColumns.DisplayName),
            enabled: (record.getValue(ItemColumns.Enabled) as boolean) ?? undefined,
            iconName: record.getFormattedValue(ItemColumns.IconName),
            iconColor: record.getFormattedValue(ItemColumns.IconColor),
            parentItemKey: record.getFormattedValue(ItemColumns.ParentKey),
            iconOnly: (record.getValue(ItemColumns.IconOnly) as boolean) ?? undefined,
            checked: (record.getValue(ItemColumns.Checked) as boolean) ?? undefined,
            visible: (record.getValue(ItemColumns.Visible) as boolean) ?? undefined,
            isHeader: (record.getValue(ItemColumns.ItemHeader) as boolean) ?? undefined,
            topDivider: (record.getValue(ItemColumns.ItemTopDivider) as boolean) ?? undefined,
            divider: (record.getValue(ItemColumns.ItemDivider) as boolean) ?? undefined,
            data: record,
        } as CanvasContextMenuItem;
    });
}

function getDummyAction(key: string): CanvasContextMenuItem {
    return {
        id: key,
        key: key,
        name: 'Item ' + key,
        iconName: 'Unknown',
    } as CanvasContextMenuItem;
}

export function getCommandsWithChildren(
    items: CanvasContextMenuItem[],
    disabled: boolean,
    onClick: (ev?: React.MouseEvent<HTMLButtonElement>, item?: IContextualMenuItem | undefined) => boolean,
): IContextualMenuItem[] {
    return items
        .filter((i) => !i.parentItemKey && i.visible !== false)
        .map((i) => {
            return getCommandBarItemProps(items, i, disabled, onClick);
        });
}

export function getCommandBarItemProps(
    items: CanvasContextMenuItem[],
    item: CanvasContextMenuItem,
    disabled: boolean,
    onClick: (ev?: React.MouseEvent<HTMLButtonElement>, item?: IContextualMenuItem | undefined) => boolean,
): IContextualMenuItem {
    const subItems = items
        .filter((i) => i.parentItemKey === item.key && i !== item && i.visible !== false)
        .map((i) => {
            return getCommandBarItemProps(items, i, disabled, onClick);
        });
    const buttonStyles = {
        root: {
            background: 'rgba(255, 255, 255,0)',
        },
    } as IButtonStyles;

    const iconProps =
        item.iconName &&
        ({
            iconName: item.iconName,
        } as IIconProps);

    // If custom icon color, find the shade slots for the hover/pressed state
    if (iconProps && item.iconColor) {
        const iconColor = item.iconColor;
        iconProps.styles = { root: { color: item.iconColor + CSS_IMPORTANT } };

        // Attempt to set a darker shade for hover/pressed colors
        const primaryColor = getColorFromString(iconColor);
        if (primaryColor) {
            const iconColorDarker = getShade(primaryColor, Shade.Shade8);
            const secondaryColorStyle = { color: iconColorDarker?.str + CSS_IMPORTANT };
            buttonStyles.iconHovered = secondaryColorStyle;
            buttonStyles.iconPressed = secondaryColorStyle;
            buttonStyles.iconExpanded = secondaryColorStyle;
            buttonStyles.iconChecked = secondaryColorStyle;
            buttonStyles.iconExpandedHovered = secondaryColorStyle;
        }
    }

    const props = {
        key: item.key,
        text: item.name,
        disabled: item.enabled === false || disabled,
        onClick: onClick,
        iconOnly: item.iconOnly === true,
        iconProps: iconProps,
        canCheck: item.checked !== undefined,
        checked: item.checked,
        data: item,
        buttonStyles: buttonStyles,
    } as IContextualMenuItem;

    if (item.isHeader === true && subItems.length > 0) {
        // Semantic Section
        props.itemType = ContextualMenuItemType.Section;
        props.sectionProps = {
            title: item.name,
            topDivider: item.topDivider,
            bottomDivider: item.divider,
            items: subItems,
        };
    } else if (item.isHeader === true) {
        // Simple Section Break
        props.itemType = ContextualMenuItemType.Header;
    } else if (item.divider === true) {
        // Divider Break
        props.itemType = ContextualMenuItemType.Divider;
    } else {
        props.itemType = ContextualMenuItemType.Normal;
        props.subMenuProps =
            subItems.length > 0
                ? {
                      items: subItems,
                  }
                : undefined;
    }
    return props;
}
