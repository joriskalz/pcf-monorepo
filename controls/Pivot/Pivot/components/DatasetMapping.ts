import { PivotLink } from './Component.types';
import { ItemColumns } from '../ManifestConstants';

export function getPivotItemsFromDataset(dataset: ComponentFramework.PropertyTypes.DataSet): PivotLink[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return [getDummyAction('1'), getDummyAction('2'), getDummyAction('3')];
    }
    const keyIndex: Record<string, number> = {};
    return dataset.sortedRecordIds.map((id) => {
        const record = dataset.records[id];
        // Prevent duplicate keys by appending the duplicate index
        let key = record.getValue(ItemColumns.Key) as string;
        if (keyIndex[key] !== undefined) {
            keyIndex[key]++;
            key += `_${keyIndex[key]}`;
        } else keyIndex[key] = 1;
        return {
            id: record.getRecordId(),
            key: key,
            name: record.getValue(ItemColumns.DisplayName) as string,
            enabled: record.getValue(ItemColumns.Enabled) as boolean,
            iconName: record.getValue(ItemColumns.IconName) as string,
            iconColor: record.getValue(ItemColumns.IconColor) as string,
            textColor: record.getValue(ItemColumns.TextColor) as string,
            parentItemKey: record.getValue(ItemColumns.ParentKey) as string,
            iconOnly: record.getValue(ItemColumns.IconOnly) as boolean,
            checked: record.getValue(ItemColumns.Checked) as boolean,
            visible: record.getValue(ItemColumns.Visible) as boolean,
            itemCount: (record.getValue(ItemColumns.ItemCount) as number) ?? undefined,
            data: record,
        } as PivotLink;
    });
}

function getDummyAction(key: string): PivotLink {
    return {
        id: key,
        key: key,
        name: 'Item ' + key,
        iconName: 'Unknown',
    } as PivotLink;
}
