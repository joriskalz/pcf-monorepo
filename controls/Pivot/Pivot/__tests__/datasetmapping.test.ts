import { getPivotItemsFromDataset } from '../components/DatasetMapping';
import { ItemColumns } from '../ManifestConstants';
import { MockDataSet, MockEntityRecord } from '../__mocks__/mock-datasets';

describe('DatasetMapping', () => {
    beforeEach(() => jest.clearAllMocks());

    it('returns correct props', () => {
        const items = [
            new MockEntityRecord('1', {
                [ItemColumns.Key]: 'item1',
                [ItemColumns.DisplayName]: 'Item 1',
                [ItemColumns.IconName]: 'Open',
                [ItemColumns.IconColor]: 'blue',
                [ItemColumns.Enabled]: true,
                [ItemColumns.IconOnly]: false,
            }),
            new MockEntityRecord('2', {
                [ItemColumns.Key]: 'item2',
                [ItemColumns.DisplayName]: 'Item 2',
                [ItemColumns.IconName]: 'New',
                [ItemColumns.IconColor]: 'blue',
                [ItemColumns.Enabled]: false,
                [ItemColumns.IconOnly]: true,
            }),
            new MockEntityRecord('3', {
                [ItemColumns.Key]: 'item3',
                [ItemColumns.DisplayName]: 'Item 3',
                [ItemColumns.IconName]: 'Save',
                [ItemColumns.IconColor]: 'green',
                [ItemColumns.Enabled]: false,
                [ItemColumns.IconOnly]: true,
            }),
            // Sub Items First Level
            new MockEntityRecord('5', {
                [ItemColumns.Key]: 'item5',
                [ItemColumns.DisplayName]: 'Item 5',
                [ItemColumns.IconName]: 'New',
                [ItemColumns.IconColor]: 'blue',
                [ItemColumns.Enabled]: false,
                [ItemColumns.IconOnly]: true,
                [ItemColumns.ParentKey]: 'commandNew',
            }),
            new MockEntityRecord('6', {
                [ItemColumns.Key]: 'item6',
                [ItemColumns.DisplayName]: 'Item 6',
                [ItemColumns.IconName]: 'Save',
                [ItemColumns.IconColor]: 'green',
                [ItemColumns.Enabled]: false,
                [ItemColumns.IconOnly]: true,
                [ItemColumns.ParentKey]: 'commandSave',
            }),
            // Sub Items Second Level
            new MockEntityRecord('7', {
                [ItemColumns.Key]: 'item7',
                [ItemColumns.DisplayName]: 'Item 7',
                [ItemColumns.IconName]: 'New',
                [ItemColumns.IconColor]: 'blue',
                [ItemColumns.Enabled]: false,
                [ItemColumns.IconOnly]: true,
                [ItemColumns.ParentKey]: 'item1',
            }),
            new MockEntityRecord('8', {
                [ItemColumns.Key]: 'item8',
                [ItemColumns.DisplayName]: 'Item 8',
                [ItemColumns.IconName]: 'New',
                [ItemColumns.IconColor]: 'blue',
                [ItemColumns.Enabled]: false,
                [ItemColumns.IconOnly]: true,
                [ItemColumns.ParentKey]: 'item2',
            }),
        ];

        const actions = getPivotItemsFromDataset(new MockDataSet(items));
        expect(actions).toMatchSnapshot();
    });

    it('removes duplicate keys', () => {
        const itemsWithDuplicates = [
            new MockEntityRecord('1', {
                [ItemColumns.Key]: 'key1',
                [ItemColumns.DisplayName]: 'Open',
            }),
            new MockEntityRecord('2', {
                [ItemColumns.Key]: 'key1',
                [ItemColumns.DisplayName]: 'Open',
            }),
            new MockEntityRecord('3', {
                [ItemColumns.Key]: 'key1',
                [ItemColumns.DisplayName]: 'Open',
            }),
            new MockEntityRecord('4', {
                [ItemColumns.Key]: '2',
                [ItemColumns.DisplayName]: 'Open',
            }),
            new MockEntityRecord('5', {
                [ItemColumns.Key]: '2',
                [ItemColumns.DisplayName]: 'Open',
            }),
            new MockEntityRecord('6', {
                [ItemColumns.Key]: 'unique3',
                [ItemColumns.DisplayName]: 'Open',
            }),
        ];

        const actions = getPivotItemsFromDataset(new MockDataSet(itemsWithDuplicates));
        expect(actions.length).toBe(6);
        expect(actions[0].key).toBe('key1');
        expect(actions[1].key).toBe('key1_2');
        expect(actions[2].key).toBe('key1_3');
        expect(actions[3].key).toBe('2');
        expect(actions[4].key).toBe('2_2');
        expect(actions[5].key).toBe('unique3');
    });
});
