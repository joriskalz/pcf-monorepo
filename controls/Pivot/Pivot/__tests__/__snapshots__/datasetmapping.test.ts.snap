// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetMapping returns correct props 1`] = `
Array [
  Object {
    "checked": undefined,
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "1",
      "values": Object {
        "ItemDisplayName": "Item 1",
        "ItemEnabled": true,
        "ItemIconColor": "blue",
        "ItemIconName": "Open",
        "ItemIconOnly": false,
        "ItemKey": "item1",
      },
    },
    "enabled": true,
    "iconColor": "blue",
    "iconName": "Open",
    "iconOnly": false,
    "id": "1",
    "itemCount": undefined,
    "key": "item1",
    "name": "Item 1",
    "parentItemKey": undefined,
    "textColor": undefined,
    "visible": undefined,
  },
  Object {
    "checked": undefined,
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "2",
      "values": Object {
        "ItemDisplayName": "Item 2",
        "ItemEnabled": false,
        "ItemIconColor": "blue",
        "ItemIconName": "New",
        "ItemIconOnly": true,
        "ItemKey": "item2",
      },
    },
    "enabled": false,
    "iconColor": "blue",
    "iconName": "New",
    "iconOnly": true,
    "id": "2",
    "itemCount": undefined,
    "key": "item2",
    "name": "Item 2",
    "parentItemKey": undefined,
    "textColor": undefined,
    "visible": undefined,
  },
  Object {
    "checked": undefined,
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "3",
      "values": Object {
        "ItemDisplayName": "Item 3",
        "ItemEnabled": false,
        "ItemIconColor": "green",
        "ItemIconName": "Save",
        "ItemIconOnly": true,
        "ItemKey": "item3",
      },
    },
    "enabled": false,
    "iconColor": "green",
    "iconName": "Save",
    "iconOnly": true,
    "id": "3",
    "itemCount": undefined,
    "key": "item3",
    "name": "Item 3",
    "parentItemKey": undefined,
    "textColor": undefined,
    "visible": undefined,
  },
  Object {
    "checked": undefined,
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "5",
      "values": Object {
        "ItemDisplayName": "Item 5",
        "ItemEnabled": false,
        "ItemIconColor": "blue",
        "ItemIconName": "New",
        "ItemIconOnly": true,
        "ItemKey": "item5",
        "ItemParentKey": "commandNew",
      },
    },
    "enabled": false,
    "iconColor": "blue",
    "iconName": "New",
    "iconOnly": true,
    "id": "5",
    "itemCount": undefined,
    "key": "item5",
    "name": "Item 5",
    "parentItemKey": "commandNew",
    "textColor": undefined,
    "visible": undefined,
  },
  Object {
    "checked": undefined,
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "6",
      "values": Object {
        "ItemDisplayName": "Item 6",
        "ItemEnabled": false,
        "ItemIconColor": "green",
        "ItemIconName": "Save",
        "ItemIconOnly": true,
        "ItemKey": "item6",
        "ItemParentKey": "commandSave",
      },
    },
    "enabled": false,
    "iconColor": "green",
    "iconName": "Save",
    "iconOnly": true,
    "id": "6",
    "itemCount": undefined,
    "key": "item6",
    "name": "Item 6",
    "parentItemKey": "commandSave",
    "textColor": undefined,
    "visible": undefined,
  },
  Object {
    "checked": undefined,
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "7",
      "values": Object {
        "ItemDisplayName": "Item 7",
        "ItemEnabled": false,
        "ItemIconColor": "blue",
        "ItemIconName": "New",
        "ItemIconOnly": true,
        "ItemKey": "item7",
        "ItemParentKey": "item1",
      },
    },
    "enabled": false,
    "iconColor": "blue",
    "iconName": "New",
    "iconOnly": true,
    "id": "7",
    "itemCount": undefined,
    "key": "item7",
    "name": "Item 7",
    "parentItemKey": "item1",
    "textColor": undefined,
    "visible": undefined,
  },
  Object {
    "checked": undefined,
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "8",
      "values": Object {
        "ItemDisplayName": "Item 8",
        "ItemEnabled": false,
        "ItemIconColor": "blue",
        "ItemIconName": "New",
        "ItemIconOnly": true,
        "ItemKey": "item8",
        "ItemParentKey": "item2",
      },
    },
    "enabled": false,
    "iconColor": "blue",
    "iconName": "New",
    "iconOnly": true,
    "id": "8",
    "itemCount": undefined,
    "key": "item8",
    "name": "Item 8",
    "parentItemKey": "item2",
    "textColor": undefined,
    "visible": undefined,
  },
]
`;
